Some text fails to translate to Simplified Chinese on Apple TV+ page

rniwa opened the pull request:
#### f5c2130755ab39b4b005c196eabaa05406f586ea
<pre>
Some text fails to translate to Simplified Chinese on Apple TV+ page
<a href="https://bugs.webkit.org/show_bug.cgi?id=272214">https://bugs.webkit.org/show_bug.cgi?id=272214</a>
&lt;<a href="https://rdar.apple.com/72939899">rdar://72939899</a>&gt;

Reviewed by Sihui Liu and Wenson Hsieh.

The bug is caused by ParagraphContentIterator::advanceIteratorNodeAndUpdateText skipping a whitespace
generated for a tab character that appears at where a line wraps. When the layout of the page changes
slightly so that this tab character appears in the middle of a line, we don&apos;t skip the same whitespace.
As a result, TextManipulationController erroneously conclude that the content has changed and reject
the translation.

Fixed the bug by tweaking TextIterator to generate a non-collapsed range in this specific case so that
ParagraphContentIterator::advanceIteratorNodeAndUpdateText generates a whitespace as expected.
We already had a code path for a space so reuse the same code path for other collapsible whitespaces.

This PR also adds TextIterator::showTreeForThis and showTree(TextIterator&amp;) / showTree(TextIterator*)
to make debugging TextIterator easier.

Tests: TextManipulation.CompleteTextManipulationDoesNotSkipTabCharacterAtLineWrap

* Source/WebCore/editing/TextIterator.cpp:
(WebCore::TextIterator::handleTextRun):
(WebCore::TextIterator::showTreeForThis const): Added.
(showTree): Added.
* Source/WebCore/editing/TextIterator.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextManipulation.mm:
(TextManipulation.CompleteTextManipulationDoesNotSkipTabCharacterAtLineWrap):

Canonical link: <a href="https://commits.webkit.org/277150@main">https://commits.webkit.org/277150@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d59dadf0457a8e9ec3dcae65c1c774b977bd98eb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46777 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25938 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49400 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49454 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42824 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49084 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30334 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23402 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38111 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47357 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22941 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40292 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19400 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20276 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41429 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4823 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43026 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41795 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51326 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21786 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18143 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45394 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23074 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44351 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/23571 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6564 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22782 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/TextManipulation.mm:
```diff
@@ -3012,6 +3012,48 @@ - (void)_webView:(WKWebView *)webView didFindTextManipulationItem:(_WKTextManipu
 
+TEST(TextManipulation, CompleteTextManipulationDoesNotSkipTabCharacterAtLineWrap)
+{
+    auto delegate = adoptNS([[TextManipulationDelegate alloc] init]);
```
Nit: I think we prefer RetainPtr to auto nowadays

rniwa replied:
ok

rniwa committed:
Some text fails to translate to Simplified Chinese on Apple TV+ page
https://bugs.webkit.org/show_bug.cgi?id=272214
<rdar://72939899>

Reviewed by Sihui Liu and Wenson Hsieh.

The bug is caused by ParagraphContentIterator::advanceIteratorNodeAndUpdateText skipping a whitespace
generated for a tab character that appears at where a line wraps. When the layout of the page changes
slightly so that this tab character appears in the middle of a line, we don't skip the same whitespace.
As a result, TextManipulationController erroneously conclude that the content has changed and reject
the translation.

Fixed the bug by tweaking TextIterator to generate a non-collapsed range in this specific case so that
ParagraphContentIterator::advanceIteratorNodeAndUpdateText generates a whitespace as expected.
We already had a code path for a space so reuse the same code path for other collapsible whitespaces.

This PR also adds TextIterator::showTreeForThis and showTree(TextIterator&) / showTree(TextIterator*)
to make debugging TextIterator easier.

Tests: TextManipulation.CompleteTextManipulationDoesNotSkipTabCharacterAtLineWrap

* Source/WebCore/editing/TextIterator.cpp:
(WebCore::TextIterator::handleTextRun):
(WebCore::TextIterator::showTreeForThis const): Added.
(showTree): Added.
* Source/WebCore/editing/TextIterator.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextManipulation.mm:
(TextManipulation.CompleteTextManipulationDoesNotSkipTabCharacterAtLineWrap):

Canonical link: https://commits.webkit.org/277150@main
